File: android/src/main/kotlin/com/school/nfcadapter/capacitor/NfcAdapterPlugin.kt
package com.school.nfcadapter.capacitor

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.school.nfcadapter.NfcAdapterModule as CoreNfcAdapterModule
import com.school.nfcadapter.api.NfcScanListener
import com.school.nfcadapter.api.NfcScannerPort
import com.school.nfcadapter.api.ReaderError
import com.school.nfcadapter.api.ReaderState

/**
 * Capacitor 3 bridge over the core NfcAdapterModule facade.
 *
 * JS surface (see /index.js):
 *   NfcAdapter.startScanning() / stopScanning() / isSupported()
 * Events via notifyListeners:
 *   "onCardScanned"        -> { uid }  — the strict uid_dec_reversed contract value
 *   "onReaderStateChanged" -> { state }
 *   "onReaderError"        -> { code, message, recoverable }
 */
@CapacitorPlugin(name = "NfcAdapter")
class NfcAdapterPlugin : Plugin(), NfcScanListener {

    private var port: NfcScannerPort? = null

    /** Core callbacks already arrive on the Android main thread (ListenerProxy). */
    private fun ensurePort(): NfcScannerPort =
        port ?: CoreNfcAdapterModule.create(context.applicationContext).also {
            it.setListener(this)
            port = it
        }

    @PluginMethod
    fun startScanning(call: PluginCall) {
        ensurePort().start()
        call.resolve()
    }

    @PluginMethod
    fun stopScanning(call: PluginCall) {
        port?.stop()
        call.resolve()
    }

    @PluginMethod
    fun isSupported(call: PluginCall) {
        call.resolve(JSObject().put("supported", ensurePort().isSupported))
    }

    override fun handleOnDestroy() {
        port?.stop()
        port = null
        super.handleOnDestroy()
    }

    // ------------------------------------------------ NfcScanListener -> JS

    override fun onCardScanned(uidDecReversed: String) {
        notifyListeners("onCardScanned", JSObject().put("uid", uidDecReversed))
    }

    override fun onReaderStateChanged(state: ReaderState) {
        notifyListeners("onReaderStateChanged", JSObject().put("state", state.name))
    }

    override fun onReaderError(error: ReaderError) {
        notifyListeners(
            "onReaderError",
            JSObject()
                .put("code", error.code.name)
                .put("message", error.message)
                .put("recoverable", error.recoverable)
        )
    }
}

File: ios/Plugin/NfcAdapterPlugin.swift
import Foundation
import Capacitor

/// Capacitor 3 bridge over CoreNfcScannerPort. Registered with the runtime by
/// the CAP_PLUGIN macro in NfcAdapterPlugin.m under the JS name "NfcAdapter".
///
/// Events (notifyListeners):
///   "onCardScanned"        -> { uid }  — the strict uid_dec_reversed contract value
///   "onReaderStateChanged" -> { state }
///   "onReaderError"        -> { code, message, recoverable }
@objc(NfcAdapterPlugin)
public class NfcAdapterPlugin: CAPPlugin, NfcScanListener {

    private var port: NfcScannerPort?

    private func ensurePort() -> NfcScannerPort {
        if let existing = port { return existing }
        // CoreNfcScannerPort self-gates via NFCTagReaderSession.readingAvailable:
        // isSupported returns false and start() reports sessionUnavailable on
        // hardware without NFC, so no separate Noop branch is needed here.
        let created = CoreNfcScannerPort()
        created.setListener(self)
        port = created
        return created
    }

    @objc func startScanning(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.ensurePort().start()
            call.resolve()
        }
    }

    @objc func stopScanning(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.port?.stop()
            call.resolve()
        }
    }

    @objc func isSupported(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve(["supported": self.ensurePort().isSupported])
        }
    }

    // ------------------------------------------------ NfcScanListener -> JS
    // CoreNfcScannerPort already delivers these on the main queue.

    public func onCardScanned(_ uidDecReversed: String) {
        notifyListeners("onCardScanned", data: ["uid": uidDecReversed])
    }

    public func onReaderStateChanged(_ state: ReaderState) {
        notifyListeners(
            "onReaderStateChanged",
            data: ["state": String(describing: state).uppercased()]
        )
    }

    public func onReaderError(_ error: ReaderError) {
        notifyListeners("onReaderError", data: [
            "code": String(describing: error.code),
            "message": error.message,
            "recoverable": error.recoverable,
        ])
    }
}
